qep-343-customized-toolbars-and-menus.md

```diff
@@ -0,0 +1,82 @@
+
+Custom toolbars can currently be created with the plugin [Customize ToolBars](https://plugins.qgis.org/plugins/CustomToolBar/). The plugin has good ratings and has been downloaded frequently, showing that there is an interest for this functionality. Limitations of this plugin are:
+
+  * the storing of the settings in a binary file outside the QGIS folder structure
+  * possible maintenance issues
+  * loading of saved configurations is not possible through the plugin GUI (less intuitive for users)
+  * it's a plugin
+
+## Proposed Solution
+
+The integration of customized toolbars within the Interface Customization dialog would be a logical follow-up to the already available GUI customization possibilities. The configuration would also be saved within the user profile QGIS > QGISCUSTOMIZATION3.ini file.
+
+Additions to the Interface Customization dialog would be:
+
+  * similarly to the attribute form settings, a drop-down that allows to use a "drag and drop designer" for the configuration of the different components of the GUI
+  * if the "drag and drop designer" is chosen, the possibility (green plus icon) to add new "containers" (similarly to adding new containers in the attribute form settings), with the choice between "menu" and "toolbar"
+  * a "provide .ini file" option to load an existing configuration
+  * an "export configuration" button opening the profile folder at QGIS > QGISCUSTOMIZATION.ini
+
+Some quality of life ideas are:
+
+  * being able to add every element from a QGIS native toolbar/menu in one click into a custom toolbar/menu
+  * being able to duplicate an already existing custom toolbar/menu
+
+
+Possible integration of a drop-down menu similar to the layer attribute form dialog:
+![](./images/qep343/drop-down-menu.png)
+
+Current interface of the Customize ToolBars plugin:
+![](./images/qep343/customize-toolbars-plugin.png)
+
+Layer attribute form dialog (drag and drop designer; add container):
+![](./images/qep343/drag-drop-designer-container.png)
+
+### Processing algorithms
+
+Currently, processing algorithms can be added to the "Processing Algorithms Toolbar", through a non-intuitive and cumbersome user experience.
+
+To replace this logic, we propose to have the possibility of integrating any processing algorithm into the new customization of toolbars and menus.
+
+### Affected Files
+
+- src/app/qgscustomization.[cpp|h]
+
+The new interface to customize toolbars and menus will be new sections in the Interface Customization dialog:
+
+![Example of new sections in the Interface Customization dialog](./images/qep343/proposal-settings.png)
+
+### Delimitation
+
+  * A restart of QGIS could be necessary to take into account the changes (as it is already the case for modifications in the Interface Customization dialog).
```

Comment:
>  A restart of QGIS could be necessary to take into account the changes (as it is already the case for modifications in the Interface Customization dialog).

This is a very big limitation. Unlike the current interface customization feature which is mainly targeted to admins creating qgis deployments, this one should be targeting average end users. Creating custom toolbars should be as simple as moving them around and toggling them on and off

Reply:
I agree that when we decide if this feature should require a restart or not, we have to include this decision in this initial QEP.